lib/core/src/Cardano/Pool/Metrics.hs

```diff
@@ -219,25 +224,45 @@ newStakePoolLayer db@DBLayer{..} nl tr = StakePoolLayer
         when (Map.null distr || Map.null prod) $ do
             computeProgress nodeTip >>= throwE . ErrMetricsIsUnsynced
 
-        perfs <- liftIO $
-            readPoolsPerformances db epochLength (nodeTip ^. #slotId)
+        if nodeEpoch == genesisEpoch
```

Comment:
I think this can lead to a race-condition! (If the node rolls back, we could end up with a shuffled epoch 1)

Could we use the epoch of `poolProductionCursor` instead? We already fail when it is `Nothing`, I believe the change should be trivial.

Also, if you think it makes sense: could we just switch `readPoolsPerformances` to use the `poolProductionCursor` as tip at the same time, such that the performance calculation use the same source of data for `number of produced blocks by pool` and `total number of slots`?

Reply:
I am not sure to follow you... `nodeEpoch` is the epoch used for the pool production and the stake distribution. If the previous calls return nothing, then we say we are `not synced`. 

> could we just switch readPoolsPerformances to use the poolProductionCursor

Hmmm. Good point.

Reply:
Consider the scenario:
1. stake-pool-worker syncs to slot `1.3`
2. node rolls back to `0.18`
3. user calls `listStakePools` without the worker yet noticing

Then: 

- It will return block-productions for epoch 1.
- The results will be sorted randomly because `nodeEpoch == 0`
- Assuming slots per epoch = 20, total number of slots for epoch 0 will be 18, and 20 for epoch 1, while perfect productions would have yielded 20 for epoch 0 and 3 for epoch 1. I.e **way** off 😩 

Hm. It would actually be comforting to fill in some some tests here.

Reply:
Hmmm. No. It will return block production for epoch 0, and it'll be sorted randomly as it should.

Reply:
Ah! Right!

But what about the reverse: chain-follower at `0.18` while the node is `2.2`.

If we compare with the node epoch, the `0.18` productions will be sorted after be performance, which is not as intended?

Reply:
No. When we fetch the node tip, the chain follower synchronizes its tip with the node. Now, if the node tip is at `1` and we don't have the data yet, then we would exit through the previous lines when checking:

```hs
        when (Map.null distr || Map.null prod) $ do
            computeProgress nodeTip >>= throwE . ErrMetricsIsUnsynced
```

The production will be `null` and, nothing get passed that.

Reply:
> When we fetch the node tip, the chain follower synchronizes its tip with the node.

Sounds like `monitorStakePools`, but not `listStakePools`?

> The production will be null and, nothing get passed that.

Ok, good.

Reply:
> When we fetch the node tip, the chain follower synchronizes its tip with the node.

That's inherent to the network layer internals actually. Remember: the network layer is stateful. On every call to `networkTip`, it updates its list of unstable block headers.

Reply:
> Remember: the network layer is stateful.

It occurred to me. But even if the unstable block headers are updated, it does not magically force the `monitorStakePools`-chain-follower to sync? The chain follower might be sleeping? 🧐

Reply:
No it does not, but the node tip you get IS the right one. Then, there might be no data yet in the database, where my second remark applies (`when Map.null ...`)